plugin/associations.go

```diff
@@ -0,0 +1,133 @@
+package plugin
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/gogo/protobuf/protoc-gen-gogo/generator"
+	gorm "github.com/infobloxopen/protoc-gen-gorm/options"
+)
+
+func (p *OrmPlugin) parseAssociations(msg *generator.Descriptor) {
+	typeName := generator.CamelCaseSlice(msg.TypeName())
+	ormable := p.getOrmable(typeName)
+	for _, field := range msg.GetField() {
+		fieldOpts := getFieldOptions(field)
+		if fieldOpts.GetDrop() {
+			continue
+		}
+		fieldName := generator.CamelCase(field.GetName())
+		fieldType, _ := p.GoType(msg, field)
+		if p.isOrmable(fieldType) {
+			if fieldOpts == nil {
+				fieldOpts = &gorm.GormFieldOptions{}
+			}
+			childOrmable := p.getOrmable(fieldType)
+			if field.IsRepeated() {
+				// if fieldOpts.GetManyToMany() != nil {
+				// }
+				p.parseHasMany(typeName, ormable, fieldName, childOrmable, fieldOpts)
+				fieldType = fmt.Sprintf("[]*%s", childOrmable.Name)
+
+			} else {
+				// if fieldOpts.GetBelongsTo() != nil {
+				// }
+				p.parseHasOne(typeName, ormable, fieldName, childOrmable, fieldOpts)
+				fieldType = fmt.Sprintf("*%s", childOrmable.Name)
+			}
+			ormable.Fields[fieldName] = &Field{Type: fieldType, GormFieldOptions: fieldOpts}
+		}
+	}
+}
+
+func (p *OrmPlugin) parseHasMany(typeName string, parent *OrmableType, fieldName string, child *OrmableType, opts *gorm.GormFieldOptions) {
+	hasMany := opts.GetHasMany()
+	if hasMany == nil {
+		hasMany = &gorm.HasManyOptions{}
+		opts.HasMany = hasMany
+	}
+	var assocKey *Field
+	var assocKeyName string
+	if assocKeyName = generator.CamelCase(hasMany.GetAssociationForeignkey()); assocKeyName == "" {
+		assocKeyName, assocKey = p.findPrimaryKey(parent)
+	} else {
+		var ok bool
+		assocKey, ok = parent.Fields[assocKeyName]
+		if !ok {
+			p.Fail("Missing", assocKeyName, "field in", parent.Name, ".")
+		}
+	}
+	hasMany.AssociationForeignkey = &assocKeyName
+	foreignKey := &Field{Type: assocKey.Type, GormFieldOptions: &gorm.GormFieldOptions{Tag: hasMany.GetForeignkeyTag()}}
+	var foreignKeyName string
+	if foreignKeyName = hasMany.GetForeignkey(); foreignKeyName == "" {
+		foreignKeyName = fmt.Sprintf(fieldName + typeName + assocKeyName)
```

Comment:
This longer generated name should prevent any FK collisions, but is there an advantage to using it over the simpler gorm standard of {Foreign Type}ID? Especially since this has to match the DB Schema

Reply:
I made it look like that at the very last moment when I heard about multi-dimensional associations. This case is listed as a [requirement](https://docs.google.com/document/d/1XLheB0qktFQbCVKANQN9MtnrCYutUYgW2_vA19YkCZ4/edit#heading=h.32x537vq59t3) from one application team. `fieldName` prefix is used to create separate foreign keys for each dimension.
Also we already have example for this in `demo_types`: 
```golang
type TypeWithIDORM struct {
	Id            uint32
	Ip            string          `gorm:"column:ip_addr"`
	SecretInt     int32           `gorm:"-"`
	Things        []*TestTypesORM `gorm:"foreignkey:ThingsTypeWithIDId;association_foreignkey:Id"`
	ANestedObject *TestTypesORM   `gorm:"foreignkey:ANestedObjectTypeWithIDId;association_foreignkey:Id"`
}
```
Now I realize that we can do it a bit more clever: say, if we have one-dimensional  association: use `typeName + assocKeyName`, if multi-dimensional: `fieldName + typeName + assocKeyName`.
Either way these are only default values and can easily be overridden.

Reply:
Regarding generated files are different.. I also see this now. The reason is that golang maps don't preserve the order of keys and fields are generated in arbitrary order every time..:) Will fix that.

Reply:
That is a good point. We also have company internal use cases for that. Would a one/multi-dimensional check be a pretty quick thing to add? I'd really like for it to default to the simpler option if the simpler option works

Reply:
Yeah, it's already in.